pkg/operator/controller/garden/garden/reconciler.go

```diff
@@ -151,11 +151,20 @@ func (r *Reconciler) patchConditionToFalse(ctx context.Context, log logr.Logger,
 
 func (r *Reconciler) reportProgress(log logr.Logger, garden *operatorv1alpha1.Garden) flow.ProgressReporter {
 	return flow.NewImmediateProgressReporter(func(ctx context.Context, stats *flow.Stats) {
-		patch := client.MergeFrom(garden.DeepCopy())
-		conditionReconciled := v1beta1helper.GetOrInitConditionWithClock(r.Clock, garden.Status.Conditions, operatorv1alpha1.GardenReconciled)
-		garden.Status.Conditions = v1beta1helper.MergeConditions(garden.Status.Conditions, v1beta1helper.UpdatedConditionWithClock(r.Clock, conditionReconciled, gardencorev1beta1.ConditionProgressing, conditionReasonPrefix(garden)+"Progressing", fmt.Sprintf("Garden operation is currently being processed (%s (%d%%)).", strings.Join(stats.Running.StringList(), ", "), stats.ProgressPercent())))
-		if err := r.RuntimeClientSet.Client().Status().Patch(ctx, garden, patch); err != nil {
-			log.Error(err, "Could not report reconciliation progress")
+		for i := 0; i < 2; i++ {
```

Comment:
Why do you use just two iterations? Why do we need to iterate at all?

Reply:
`garden-garden` and `garden-care` controllers are both updating the `conditions` of a garden. 
Strategic merge patches are not supported for CRDs, so we have to merge with an optimistic lock. This could fail in `garden-garden` controller because it is working on a copy of the `garden` which might be outdated.
In this case ProgressReporter gets the latest version of the garden and tries patching again. I limited the numbers of retries, that there is no infinite retry loop in case of a stale cache. However, I increased the number of iterations to three.

Reply:
Will the get call actually retrieve a new copy or is it just getting the cached one? In other words, is the client used for getting the new copy using a cache or not?

Reply:
It should use a cache. Care controller is using the [client of the manager](https://github.com/gardener/gardener/blob/79c163586f79e667cf9516401e51cb8c1360b055/pkg/operator/controller/garden/care/add.go#L52) which does [not disable any cache](https://github.com/gardener/gardener/blob/79c163586f79e667cf9516401e51cb8c1360b055/cmd/gardener-operator/app/app.go#L122-L143).

Reply:
wrong controller 😅  However, it is similar for [Garden controller](https://github.com/gardener/gardener/blob/79c163586f79e667cf9516401e51cb8c1360b055/pkg/operator/controller/garden/garden/add.go#L43-L48).

Reply:
If it is uses a cache how does this ensure that we actually get a new copy? Would we not just try three times with the cached version, i.e. if the initial patch fails it would simply fail three times with the same data?

Reply:
That's part of [controller-runtime magic](https://github.com/gardener/gardener/blob/master/docs/development/kubernetes-clients.md#controller-runtime-caches) 😄  It watches the resources and responds to `Get` requests from cache.